udf-compiler/src/main/scala/com/nvidia/spark/udf/Instruction.scala

```diff
@@ -452,6 +474,91 @@ case class Instruction(opcode: Int, operand: Int, instructionStr: String) extend
     }
   }
 
+  private def arrayOp(lambdaReflection: LambdaReflection,
+      methodName: String, args: List[Expression]): Expression = {
+    // Make sure that the objref is scala.math.package$.
+    args.head match {
+      case IntegerLiteral(index) =>
+        if (!lambdaReflection.lookupField(index.asInstanceOf[Int])
+            .getType.getName.equals("scala.Array$")) {
+          throw new SparkException("Unsupported array function objref: " + args.head)
+        }
+      case _ =>
+        throw new SparkException("Unsupported array function objref: " + args.head)
+    }
+    // Translate to Catalyst
+    methodName match {
+      case "empty" =>
+        if (args.last.isInstanceOf[Repr.ClassTag[_]]) {
+          val classTag = args.last.asInstanceOf[Repr.ClassTag[_]].classTag
+          if (classTag == scala.reflect.ClassTag.Boolean) {
+            Literal(Array.empty[Boolean])
+          } else if (classTag == scala.reflect.ClassTag.Byte) {
+            Literal(Array.empty[Byte])
+          } else if (classTag == scala.reflect.ClassTag.Short) {
+            Literal(Array.empty[Short])
+          } else if (classTag == scala.reflect.ClassTag.Int) {
+            Literal(Array.empty[Int])
+          } else if (classTag == scala.reflect.ClassTag.Long) {
+            Literal(Array.empty[Long])
+          } else if (classTag == scala.reflect.ClassTag.Float) {
+            Literal(Array.empty[Float])
+          } else if (classTag == scala.reflect.ClassTag.Double) {
+            Literal(Array.empty[Double])
+          } else if (classTag == scala.reflect.ClassTag("".getClass)) {
+            Literal(Array.empty[String])
+          } else {
+            throw new SparkException("Unsupported data type for Array.empty")
+          }
+        } else {
+          throw new SparkException("Unexpected argument for Array.empty")
+        }
+      case _ => throw new SparkException("Unsupported array function: " + methodName)
+    }
+  }
+
+  private def classTagOp(lambdaReflection: LambdaReflection,
+      methodName: String, args: List[Expression]): Expression = {
+    // Make sure that the objref is scala.math.package$.
+    args.head match {
+      case IntegerLiteral(index) =>
+        if (!lambdaReflection.lookupField(index.asInstanceOf[Int])
+            .getType.getName.equals("scala.reflect.ClassTag$")) {
+          throw new SparkException("Unsupported classTag function objref: " + args.head)
+        }
+      case _ =>
+        throw new SparkException("Unsupported classTag function objref: " + args.head)
+    }
+    // Translate to Catalyst
+    methodName match {
```

Comment:
would be nice to add a comment on how the `ClassTag` code/bytecode is found in the UDFs.

Reply:
you want us to explain why Scala compiler generates `ClassTag` code?
I don't know if we want to document the behavior of Scala compiler in our code.

Reply:
I am not asking for all the behavior to be documented. I think that if we check for something that looks pretty far from a catalyst expression, yet it becomes one, it would be really great to have a small comment stating why.

Reply:
The references to `ClassTag` also come from `() => Array.empty[T]` like `checkcast`.

```
         0: getstatic     #52                 // Field scala/Array$.MODULE$:Lscala/Array$;
         3: getstatic     #57                 // Field scala/reflect/ClassTag$.MODULE$:Lscala/reflect/ClassTag$;
         6: invokevirtual #61                 // Method scala/reflect/ClassTag$.Int:()Lscala/reflect/ClassTag;
         9: invokevirtual #65                 // Method scala/Array$.empty:(Lscala/reflect/ClassTag;)Ljava/lang/Object;
        12: checkcast     #67                 // class "[I"
        15: areturn
```

Do you still need comments about `ClassTag`?

Reply:
This is good enough for me, I don't think I need this in the code.